### PR TITLE
samples: Add sample snippets for count aggregations

### PR DESCRIPTION
**Pull request by jainsahab**
This PR contains the samples for Query count feature.

### CODE REVIEW

File: samples/snippets/src/main/java/com/example/datastore/AggregationQuerySample.java [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    AggregationResult limitQueryResult = Iterables.getOnlyElement(
+        datastore.runAggregation(allCandidatesCountQuery));
+
+    System.out.printf("We have at least %d candidates", limitQueryResult.get("at_least")); // 2
+
+    // [END datastore_count_aggregation_query_with_limit]
+  }
+
+  public void aggregationQueryAndCountAggregationWithOrderBy() {
+    // [START datastore_count_aggregation_query_with_limit]
```
Comment (jlara310) [inline samples/snippets/src/main/java/com/example/datastore/AggregationQuerySample.java:119]: This snippet tag is the same as the one for the previous snippet. Please update.

Comment (jainsahab) [reply samples/snippets/src/main/java/com/example/datastore/AggregationQuerySample.java]: Thanks @jlara310 for pointing this out, Fixed as part of https://github.com/googleapis/java-datastore/pull/871/commits/4a9c7baaace44aa617682c5f1e9a6fa0b50d47bf.

File: samples/snippets/src/main/java/com/example/datastore/AggregationQuerySample.java [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+  public void aggregationQueryAndCountAggregationOnKind() {
+    // [START datastore_count_aggregation_query_on_kind]
+
+    // Instantiates a client
+    Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+
+    // The kind for the new entity
+    String kind = "Candidate";
+
+    Key candidate1Key = datastore.newKeyFactory().setKind(kind).newKey("candidate1");
```
Comment (kolea2) [inline samples/snippets/src/main/java/com/example/datastore/AggregationQuerySample.java:48]: can we keep this domain similar to already existing samples (using Tasks)?

**Review by kolea2 — CHANGES_REQUESTED** [review]

**Comment by Sita04** [comment]
Can we split the functions (with region tags) into individual java files?
It is recommended to have just one region tag / usecase per file.

File: samples/snippets/src/main/java/com/example/datastore/AggregationQuerySample.java (hunk 2 above)
Comment (jainsahab) [reply samples/snippets/src/main/java/com/example/datastore/AggregationQuerySample.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/871/commits/7d66848eae0b3c2cf4cbb5b776335859bef134e6.

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.StructuredQuery.PropertyFilter;
+import com.google.cloud.datastore.aggregation.Aggregation;
+import com.google.common.collect.Iterables;
+
+public class CountAggregationInTransaction {
+
+  public static void invoke() {
+    // [START datastore_count_aggregation_query_in_transaction]
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java:34]: The start region tag should be placed right before the import statements and the end region tag should be the last line in the file.
Kindly move them accordingly. Same comment to all the files.

File: samples/snippets/src/test/java/com/example/datastore/AggregationQuerySampleTestIT.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    CountAggregationOnKind.invoke();
+
+    systemsOutRule.assertContains("Total tasks count is 3");
+    systemsOutRule.assertContains("Total tasks (accessible from default alias) is 3");
+  }
+
+  @Test
+  public void testAggregationQueryAndCountAggregationWithLimitSample() {
+    CountAggregationWithLimit.invoke();
+
```
Comment (Sita04) [inline samples/snippets/src/test/java/com/example/datastore/AggregationQuerySampleTestIT.java:64]: nit: remove empty line

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithOrderBy.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    // Save all the tasks
+    datastore.put(
+        Entity.newBuilder(task1Key).set("done", true).set("priority", 1).build(),
+        Entity.newBuilder(task2Key).set("done", false).build(), // no priority specified
+        Entity.newBuilder(task3Key).set("done", true).set("priority", 2).build());
+
+    EntityQuery selectAllTasks =
+        Query.newEntityQueryBuilder()
+            .setKind(kind)
+            .addOrderBy(OrderBy.asc("priority")) // OrderBy acts as an existence filter
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithOrderBy.java:54]: Move the comment at the end of the line to top. (ie line 54)

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithOrderBy.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    String kind = "Task";
+
+    Key task1Key = datastore.newKeyFactory().setKind(kind).newKey("task1");
+    Key task2Key = datastore.newKeyFactory().setKind(kind).newKey("task2");
+    Key task3Key = datastore.newKeyFactory().setKind(kind).newKey("task3");
+
+    // Save all the tasks
+    datastore.put(
+        Entity.newBuilder(task1Key).set("done", true).set("priority", 1).build(),
+        Entity.newBuilder(task2Key).set("done", false).build(), // no priority specified
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithOrderBy.java:48]:
```suggestion
        // Priority not specified.
        Entity.newBuilder(task2Key).set("done", false).build(),
```

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithLimit.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    Key task2Key = datastore.newKeyFactory().setKind(kind).newKey("task2");
+    Key task3Key = datastore.newKeyFactory().setKind(kind).newKey("task3");
+
+    // Save all the tasks
+    datastore.put(
+        Entity.newBuilder(task1Key).set("done", true).build(),
+        Entity.newBuilder(task2Key).set("done", false).build(),
+        Entity.newBuilder(task3Key).set("done", true).build());
+
+    EntityQuery selectAllTasks = Query.newEntityQueryBuilder().setKind(kind).setLimit(2).build();
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithLimit.java:50]:
```suggestion
    EntityQuery selectAllTasks = Query.newEntityQueryBuilder()
        .setKind(kind)
        .setLimit(2)
        .build();
```

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    Key task2Key = datastore.newKeyFactory().setKind(kind).newKey("task2");
+    Key task3Key = datastore.newKeyFactory().setKind(kind).newKey("task3");
+
+    // Save all the tasks
+    datastore.put(
+        Entity.newBuilder(task1Key).set("done", true).build(),
+        Entity.newBuilder(task2Key).set("done", false).build(),
+        Entity.newBuilder(task3Key).set("done", true).build());
+
+    GqlQuery<?> selectAllTasks =
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java:50]:
```suggestion
    // Create a GQL query to get the count of all tasks.
    GqlQuery<?> selectAllTasks =
```

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        Query.newGqlQueryBuilder(
+                "AGGREGATE COUNT(*) AS total_count, COUNT_UP_TO(2) AS count_with_limit "
+                    + "OVER (SELECT * FROM Task)")
+            .setAllowLiteral(true)
+            .build();
+    // Creating an aggregation query to get the count of all tasks
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java:56]:
```suggestion
    // Create the aggregation query builder and set the query.
```

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        Query.newAggregationQueryBuilder().over(selectAllTasks).build();
+    // Executing aggregation query
+    AggregationResult allTasksCountQueryResult =
+        Iterables.getOnlyElement(datastore.runAggregation(allTasksCountQuery));
+
+    System.out.printf(
+        "We have at least %d tasks", allTasksCountQueryResult.get("count_with_limit")); // 2
+    System.out.printf("Total tasks count is %d", allTasksCountQueryResult.get("total_count")); // 3
+
+    GqlQuery<?> completedTasks =
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java:67]:
```suggestion
    // Create a query to get the count of all completed tasks.
    GqlQuery<?> completedTasks =
```

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        Query.newGqlQueryBuilder(
+                "AGGREGATE COUNT(*) AS total_completed_count "
+                    + "OVER (SELECT * FROM Task WHERE done = true)")
+            .setAllowLiteral(true)
+            .build();
+    // Creating an aggregation query to get the count of all completed tasks
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationWithGqlQuery.java:73]:
```suggestion
    // Create the aggregation query builder and set the query.
```

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+    // Save all the tasks
+    datastore.put(
+        Entity.newBuilder(task1Key).set("owner", "john").build(),
+        Entity.newBuilder(task2Key).set("owner", "john").build());
+
+    // Using transactions to maintain consistent application state.
+    datastore.runInTransaction(
+        (TransactionCallable<Void>)
+            transaction -> {
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java:53]: Consider adding more comments within the loop construct.

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import com.google.cloud.datastore.StructuredQuery.PropertyFilter;
+import com.google.cloud.datastore.aggregation.Aggregation;
+import com.google.common.collect.Iterables;
+
+public class CountAggregationInTransaction {
+
+  public static void invoke() {
+    // [START datastore_count_aggregation_query_in_transaction]
+
+    // Instantiates a client
```
Comment (Sita04) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java:36]: Global nit across files: add fullstops (.) at the end of each comment.

**Review by Sita04 — CHANGES_REQUESTED** [review]
Thanks for the samples. Left a few comments.

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java (hunk 3 above)
Comment (jainsahab) [reply samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/871/commits/7189f80936c2aaac76204c3848646cc38872dd28.

File: samples/snippets/src/test/java/com/example/datastore/AggregationQuerySampleTestIT.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    KeyQuery allKeysQuery = Query.newKeyQueryBuilder().setKind("Task").build();
+    QueryResults<Key> allKeys = datastore.run(allKeysQuery);
+    Key[] keysToDelete = ImmutableList.copyOf(allKeys).toArray(new Key[0]);
+    datastore.delete(keysToDelete);
+  }
+
+  @Test
+  public void testAggregationQueryAndCountAggregationSample() {
+    CountAggregationOnKind.invoke();
+
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply samples/snippets/src/test/java/com/example/datastore/AggregationQuerySampleTestIT.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/871/commits/350ab615f353a32d3a7a12648f66d9dc91ed0ffa.

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java (hunk 12 above)
Comment (jainsahab) [reply samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java:52]: Fixed as part of https://github.com/googleapis/java-datastore/pull/871/commits/abd2b953ac6f8cfdbffd659117cb8d04616539bc.

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java (hunk 13 above)
Comment (jainsahab) [reply samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationInTransaction.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/871/commits/1adf105b9ec1d980a9ce453ec9213160e851663d.

File: samples/snippets/src/test/java/com/example/datastore/AggregationQuerySampleTestIT.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+public class AggregationQuerySampleTestIT {
+
+  private final Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+
+  @Rule public final SystemsOutRule systemsOutRule = new SystemsOutRule();
+
+  @Before
+  public void setUp() throws Exception {
+    KeyQuery allKeysQuery = Query.newKeyQueryBuilder().setKind("Task").build();
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply samples/snippets/src/test/java/com/example/datastore/AggregationQuerySampleTestIT.java:49]: Fixed as part of https://github.com/googleapis/java-datastore/pull/871/commits/350ab615f353a32d3a7a12648f66d9dc91ed0ffa.

File: samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationOnKind.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        Entity.newBuilder(task1Key).set("done", true).build(),
+        Entity.newBuilder(task2Key).set("done", false).build(),
+        Entity.newBuilder(task3Key).set("done", true).build());
+
+    EntityQuery selectAllTasks = Query.newEntityQueryBuilder().setKind(kind).build();
+    // Creating an aggregation query to get the count of all tasks.
+    AggregationQuery allTasksCountQuery =
+        Query.newAggregationQueryBuilder()
+            .over(selectAllTasks)
+            .addAggregation(Aggregation.count())
```
Comment (kolea2) [inline samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationOnKind.java:55]: is this meant to be added twice (one with alias, one without?)

Comment (jainsahab) [reply samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationOnKind.java:55]: Yes, to showcase that aggregations can have a custom alias.

Comment (kolea2) [reply samples/snippets/src/main/java/com/example/datastore/aggregation/CountAggregationOnKind.java:55]: calling count twice in the same query seems confusing to me, could we make this two separate examples?

**Review by kolea2 — APPROVED** [review]

**Review by Sita04 — APPROVED** [review]

**Pull request merged**